File: app_manager/appmanager/src/main/java/app/editors/manager/managers/retrofit/ProgressRequestBody.java
package app.editors.manager.managers.retrofit;

import android.content.ContentResolver;
import android.content.Context;
import android.net.Uri;

import java.io.IOException;
import java.io.InputStream;

import lib.toolkit.base.managers.utils.ContentResolverUtils;
import okhttp3.MediaType;
import okhttp3.RequestBody;
import okio.BufferedSink;

public class ProgressRequestBody extends RequestBody {

    private static final int DEFAULT_BUFFER_SIZE = 2048;

    public interface OnUploadCallbacks {
        void onProgressUpdate(long total, long progress);
    }

    private volatile boolean mIsCanceled;
    private final Context mContext;
    private final ContentResolver mContentResolver;
    private final Uri mUri;
    private final String mContentType;
    private final long mTotalSize;
    private OnUploadCallbacks mOnUploadCallbacks;

    public ProgressRequestBody(final Context context, final Uri uri) {
        mIsCanceled = false;
        mContext = context;
        mContentResolver = mContext.getContentResolver();
        mUri = uri;
        mTotalSize = ContentResolverUtils.getSize(mContext, mUri);
        mContentType = ContentResolverUtils.getMimeType(mContext, mUri);
    }

    @Override
    public MediaType contentType() {
        return MediaType.parse(mContentType);
    }

    @Override
    public long contentLength() {
        return mTotalSize;
    }

    @Override
    public void writeTo(BufferedSink sink) throws IOException {
        final byte[] buffer = new byte[DEFAULT_BUFFER_SIZE];

        try (InputStream inputStream = mContentResolver.openInputStream(mUri)) {
            int countBytes;
            long totalBytes = 0;
            while ((countBytes = inputStream.read(buffer)) != -1 && !mIsCanceled) {
                totalBytes += countBytes;
                sink.write(buffer, 0, countBytes);
                if (mOnUploadCallbacks != null) {
                    mOnUploadCallbacks.onProgressUpdate(mTotalSize, totalBytes);
                }
            }
        }
    }

    public void cancel() {
        mIsCanceled = true;
    }

    public void setOnUploadCallbacks(OnUploadCallbacks onUploadCallbacks) {
        mOnUploadCallbacks = onUploadCallbacks;
    }

}

File: app_manager/core/src/main/java/app/documents/core/manager/ProgressRequestBody.kt
package app.documents.core.manager

import android.content.ContentResolver
import android.content.Context
import android.net.Uri
import lib.toolkit.base.managers.utils.ContentResolverUtils
import okhttp3.MediaType
import okhttp3.RequestBody
import okio.BufferedSink
import java.io.IOException

class ProgressRequestBody(context: Context, val uri: Uri) : RequestBody() {

    @Volatile
    private var isCanceled = false

    private val contentResolver: ContentResolver = context.contentResolver
    private val contentType: String = ContentResolverUtils.getMimeType(context, uri)
    private val totalSize: Long = ContentResolverUtils.getSize(context, uri)
    private var onUploadCallbacks: ((total: Long, progress: Long) -> Unit)?  = null

    override fun contentType(): MediaType? {
        return MediaType.parse(contentType)
    }

    override fun contentLength(): Long {
        return totalSize
    }

    @Throws(IOException::class)
    override fun writeTo(sink: BufferedSink) {
        val buffer = ByteArray(DEFAULT_BUFFER_SIZE)
        contentResolver.openInputStream(uri).use { inputStream ->
            var countBytes: Int
            var totalBytes: Long = 0
            while (inputStream!!.read(buffer).also { countBytes = it } != -1 && !isCanceled) {
                totalBytes += countBytes.toLong()
                sink.write(buffer, 0, countBytes)
                onUploadCallbacks?.invoke(totalSize, totalBytes)
            }
        }
    }

    fun cancel() {
        isCanceled = true
    }

    fun setOnUploadCallbacks(listener: ((total: Long, progress: Long) -> Unit)?  = null) {
        this.onUploadCallbacks = listener
    }

    companion object {
        private const val DEFAULT_BUFFER_SIZE = 2048
    }


}